engine/verification/assigner/engine.go

```diff
@@ -163,15 +187,15 @@ func (e *Engine) ProcessFinalizedBlock(block *flow.Block) {
 
 	log.Debug().Msg("new finalized block arrived")
 
+	e.metrics.OnFinalizedBlockReceived()
+
 	err := e.indexer.Index(block.Payload.Receipts)
 	if err != nil {
 		// TODO: consider aborting the process
 		log.Error().Err(err).Msg("could not index receipts for block")
 	}
 
-	for _, receipt := range block.Payload.Receipts {
-		e.handleExecutionReceipt(receipt, blockID)
-	}
+	e.handleExecutionReceiptsWithTracing(ctx, block.Payload.Receipts, blockID)
```

Comment:
Are we going to let `handleExecutionReceipt` return `(chunks, error)` in order to chain the log with context?

Reply:
We are currently logging everything internally in `handleExecutionReceipt`. Per a [comment on a previous PR](https://github.com/onflow/flow-go/pull/427#discussion_r585139364), I tried returning values and logging it externally, but as I explained in that comment, I found it degrading the readability and maintainability of the code. 

> to chain the log with context

would you please explain it further, not sure how to do that and what are the advantages.

Reply:
The idea is to reuse the same logger and break the big processing steps into smaller ones, and make it clear what's the output of each step, and be able to log all the context:

```
func (e *Engine) ProcessFinalizedBlock(block) {
  defer e.blockConsumerNotifier.Notify(blockID)
  log := e.log().Hex("block_id").Int("block_height")

  receipts, err := e.findReceiptsFromBlockPayload(block.ID())
  if err != nil {
    log.Fatal().Err(err).Msg("...")
  }

  log = log.Int("receipts", len(receipts))

  mychunks, err := e.findMyChunks(blockID, receipts)
  if err {
    log.Fatal().Err(err).Msg(...)
  }  

  log = log.Int("chunks", len(mychunks))

  for _, chunk := range mychunks {
    err := e.processChunk(chunk)
    if err != nil {
      log.Fatal().Err(err).Msg(..)
    }
  }
  
 log.Info().Msg("finish processing block")
}
```

Reply:
Refactored mostly in https://github.com/onflow/flow-go/pull/466/commits/6d7caf32aeb33aabb9b7f94d20ae2603b478b621 with some small hammering in subsequent commits.